### DISCUSSION TITLE
SSD metric for affine alignment

### DISCUSSION THREAD

**Discussion by GFleishman** [dipy/dipy#2187] [Registration] [answered]
I wasn't sure if there was a more appropriate place for this question.
Has anyone implemented a SSDMetric like object appropriate for use in affine registration?
I'm familiar with `align.metrics.SSDMetric`, but as far as I can tell, its API is not consistent with the requirements of `align.imaffine.AffineRegistration`.
Of course there is also `align.imaffine.MutualInformationMetric` but for my application, I need to very quickly affine align a very large number of 3D images and mutual information is just too slow. My images were all collected on the same instrument, and I've validated SSD as a suitable metric for affine alignment using other packages, though these are unfortunately all compiled C++ packages and thus don't integrate well with the remainder of my python software stack.
If there really is nothing like this in dipy, I would consider building it myself, but I don't want to start if something is already there.

**Comment by skoudoro** [comment, marked as answer]
Hi @GFleishman,

Thank you for asking. we have an implementation of SSD in one of my branches. It just needs cleaning and redesign because there is a lot of duplicated code (the current code is ugly, sorry for that)

You can look:
- [here](https://github.com/skoudoro/dipy/blob/ssd/dipy/align/imaffine.py#L795) and [here](https://github.com/skoudoro/dipy/blob/ssd/dipy/align/ssdmetric.pyx) for the SSD code
- [here](https://github.com/skoudoro/dipy/blob/ssd/doc/examples/affine_registration_ssd.py) for an example.

Do not hesitate if you have any questions and I will appreciate any feedback. 

> its API is not consistent with the requirements of align.imaffine.AffineRegistration.

Yes, we have a lot of work on this part. This is on our TODO list
